### PR TITLE
cpu: atmega_common, msp430: misc libc fixes for fmt

### PR DESCRIPTION
**Pull request by kaspar030**
AVR & msp430 don't have write().

This PR makes fmt use fwrite() on AVR, and implements write() using uart for the msp430s.
Also includes some related libc header fixes.

Without this, fmt will not compile on those platforms.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Nice. Maybe you can add a simple application to tests that actually compiles `fmt` so that we will notice these kind of problems earlier the next time?

**Comment by kaspar030** [comment]
@OlegHahm Yep... When I get to it I'll write unittests...

**Comment by haukepetersen** [comment]
can confirm that `z1` and `arduino-mega1560` build fine with this PR, but not without it..

**Comment by kaspar030** [comment]
@OlegHahm A unittest seemed useless for a void function that would have to ask the user for successful printing, so I've added a simple test that just prints something. It doesn't build on master for some platforms, but does with this PR.

**Comment by OlegHahm** [comment]
That's actually what I had in mind. Let's see if Travis is happy.

**Comment by kaspar030** [comment]
(the added unistd.h clashes with the sys/posix one ... argl...)

**Comment by kaspar030** [comment]
I decided to fix the posix file clashes in another PR: #4321.

**Comment by haukepetersen** [comment]
@kaspar030: is this PR still waiting on any other PR (flag is set)? And what needs to be done here? In the headers are some doxygen comments missing, otherwise I think its fine. Would you mind to fix this and rebase and squash?

**Comment by kaspar030** [comment]
@haukepetersen This still needs #4321, which is just awaiting ACK. Will fix doxygen in the meantime...

**Comment by kaspar030** [comment]
- rebased
- not waiting for other PR anymore, #4321 is merged

wouldn't squash further, the changes are not related.
and i don't document C headers.

File: cpu/msp430fxyz/msp430_stdio.c [hunk 1]
```diff
@@ -40,3 +43,17 @@ int putchar(int c)
...
     return 1;
 }
+
+/**
+ * @brief   Write nbyte characters to the STDIO UART interface
+ */
+ssize_t write(int fildes, const void *buf, size_t nbyte)
+{
+    if (fildes == STDOUT_FILENO) {
+        uart_stdio_write(buf, nbyte);
```
Comment (Kijewski) [inline cpu/msp430fxyz/msp430_stdio.c]: Why don't you `return uart_stdio_write(buf, nbyte);`?

File: sys/fmt/fmt.c [hunk 2]
```diff
@@ -137,6 +137,10 @@ uint32_t scn_u32_dec(const char *str, size_t n)
 
 void print(const char *s, size_t n)
 {
+#ifdef __WITH_AVRLIBC__
+    /* AVR's libc doesn't offer write(), so use fwrite() instead */
+    fwrite(s, n, 1, stdout);
```
Comment (Kijewski) [inline sys/fmt/fmt.c]: Shouldn't this be in the loop as well?

File: cpu/msp430fxyz/msp430_stdio.c (hunk 1 above)
Comment (kaspar030) [inline cpu/msp430fxyz/msp430_stdio.c]: ```uart_stdio_write()``` doesn't return anything, as writing to uart doesn't fail on any of our platforms.

File: sys/fmt/fmt.c (hunk 2 above)
Comment (kaspar030) [inline sys/fmt/fmt.c]: I thought fwrite abstracts system-call induced half-writes and only returns 0 (instead of 1 for 1 member of size n) on eof or a hard error. No idea if they can happen on stdout on AVR.

File: cpu/msp430fxyz/msp430_stdio.c (hunk 1 above)
Comment (Kijewski) [inline cpu/msp430fxyz/msp430_stdio.c]: But it does: http://riot-os.org/api/group__sys__uart__stdio.html#gac92e7c3117f2a8398d4be7293259e022

Comment (kaspar030) [inline cpu/msp430fxyz/msp430_stdio.c]: You're right, sorry. Now I remember, ```uart_stdio_write()``` just returns the len parameter, as ```uart_write()``` is ```void(*)()```, so I thought it saves some .text ignoring the (useless) return value.

File: sys/fmt/fmt.c (hunk 2 above)
Comment (OlegHahm) [inline sys/fmt/fmt.c]: Can you rephrase please? I don't understand this.

Comment (kaspar030) [inline sys/fmt/fmt.c]: Well, write() is usually a system call, while fwrite() is a libc layer on top of that.

write() might fail for many reasons, like interruption by signal, or the system might decide to only write half of the data as it pleases.

fwrite() uses write(), but usually adds some buffering. Because of that, it has to deal with incomplete write() calls internally, as the user of fwrite cannot know how e.g., the caching is implemented.

For that reason I assume that fwrite usually succeeds unless the underlying file has a hard error (e.g., it gets closed, at disk full, or there's a hardware failure), which I also assume won't happen for the AVR's serial port.

File: cpu/msp430fxyz/msp430_stdio.c (hunk 1 above)
Comment (kaspar030) [inline cpu/msp430fxyz/msp430_stdio.c]: Do you want me to change this?

File: sys/fmt/fmt.c (hunk 2 above)
Comment (OlegHahm) [inline sys/fmt/fmt.c]: Thanks for the explanation. Makes sense to me and I guess even if the AVR's serial port fails, looping and retrying won't do any good here.

File: cpu/msp430fxyz/msp430_stdio.c (hunk 1 above)
Comment (kaspar030) [inline cpu/msp430fxyz/msp430_stdio.c]: - addressed, and it actually saved 10b code.

**Comment by kaspar030** [comment]
- squashed

**Comment by OlegHahm** [comment]
ACK

**Comment by kaspar030** [comment]
Travis is happy -> go!
